rest/blip_client_test.go

```diff
@@ -65,19 +70,153 @@ type BlipTesterClient struct {
 	nonCollectionAwareClient *BlipTesterCollectionClient
 }
 
+func (c *BlipTesterCollectionClient) getClientDocForSeq(seq clientSeq) (*clientDoc, bool) {
+	c.seqLock.RLock()
+	defer c.seqLock.RUnlock()
+	doc, ok := c._seqStore[seq]
+	return doc, ok
+}
+
+// OneShotDocsSince is an iterator that yields client sequence and document pairs that are newer than the given since value.
+func (c *BlipTesterCollectionClient) OneShotDocsSince(since clientSeq) iter.Seq2[clientSeq, *clientDoc] {
+	return func(yield func(clientSeq, *clientDoc) bool) {
+		c.seqLock.Lock()
+		seqLast := c._seqLast
+		for c._seqLast <= since {
+			// block until new seq
+			c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - waiting for new sequence", since, c._seqLast)
+			c._seqCond.Wait()
+			seqLast = c._seqLast
+			c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - woke up", since, c._seqLast)
+		}
+		c.seqLock.Unlock()
+		c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - iterating", since, seqLast)
+		for seq := since; seq <= seqLast; seq++ {
+			doc, ok := c.getClientDocForSeq(seq)
+			// filter non-latest entries in cases where we haven't pruned _seqStore
+			if !ok {
+				continue
+			} else if doc.latestSeq != seq {
+				// this entry should be cleaned up from _seqStore?
+				base.AssertfCtx(context.TODO(), "seq %d found in _seqStore but latestSeq for doc %d - this should've been pruned out!", seq, doc.latestSeq)
+				continue
+			}
+			if !yield(seq, doc) {
+				c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - stopping iteration", since, c._seqLast)
+				return
+			}
+		}
+		c.TB().Logf("OneShotDocsSince: since=%d, _seqLast=%d - done", since, c._seqLast)
+	}
+}
+
+func (c *BlipTesterCollectionClient) docsSince(ctx context.Context, since clientSeq, continuous bool) chan *clientDoc {
+	ch := make(chan *clientDoc)
+	go func() {
+		sinceVal := since
+		for {
+			c.TB().Logf("docsSince: sinceVal=%d", sinceVal)
+			for _, doc := range c.OneShotDocsSince(sinceVal) {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				c.TB().Logf("sending doc %q", doc.id)
+				ch <- doc
+				sinceVal = doc.latestSeq
+			}
+			if !continuous {
+				c.TB().Logf("opts.Continuous=false, breaking changes loop")
+				break
+			}
+		}
+	}()
+	return ch
+}
+
+type clientSeq uint64
+
+// clientDocRev represents a revision of a document stored on this client, including any metadata associated with this specific revision.
+type clientDocRev struct {
+	clientSeq     clientSeq
+	version       DocVersion
+	parentVersion *DocVersion
+	body          []byte
+	isDelete      bool
+	isNoRev       bool
+	message       *blip.Message // rev or norev message associated with this revision
```

Comment:
I'm confused by the presence of the message property. If this was a pull replication, this is storing the rev messages. What's the advantage of storing the rev messages with the other information?

Reply:
so tests can assert on the replication protocol behaviour specifically, rather than the eventual outcome (e.g. identifying whether this rev was sent as a delta, and introspecting what message properties it had)